### ISSUE TITLE
Running demos/sslecho yields certificate verification error

### ISSUE THREAD

**Issue by exitc0de** [openssl/openssl#19639]
Running the sslecho demo code from this repo (https://github.com/openssl/openssl/tree/master/demos/sslecho), with both the provided key.pem and cert.pem as well as self-generated, self-signed certificates and keys yields the following error when running the client side: 

```
TCP connection to server successful
SSL connection to server failed

80457F0001000000:error:0A000086:SSL routines:tls_post_process_server_certificate:certificate verify failed:ssl/statem/statem_clnt.c:1890:
sslecho exiting
```

I generated my own key and cert with both the provided command and this one:
```
openssl req -config localhost.conf -new -x509 -sha256 -newkey rsa:4096 -nodes -keyout key.pem -days 3650 -out cert.pem
```

where localhostl.conf specifies localhost and 127.0.0.1 as subjectAltNames and commonName = localhost, basicConstraints = CA:FALSE and keyUsage = digitalSignature, keyEncipherment, keyCertSign.

I am using OpenSSL 3.0.7 installed in homebrew on OS X 12.6 Monteray on an Apple M1 Pro, and have verified that the certificate is saved in my my trusted certs folder, `/opt/homebrew/etc/openssl@3/certs`.

Strangely, running the server-side of the code and then connecting a client using `s_client` works perfectly:

```
openssl s_client -showcerts -connect localhost:4433
```

Output is as follows:

```
CONNECTED(00000005)
Can't use SSL_get_servername
depth=0 C = GB, ST = London, L = London, O = My Company, CN = localhost, emailAddress = email@email.com
verify return:1
---
Certificate chain
 0 s:C = GB, ST = London, L = London, O = My Company, CN = localhost, emailAddress = email@email.com
   i:C = GB, ST = London, L = London, O = My Company, CN = localhost, emailAddress = email@email.com
   a:PKEY: rsaEncryption, 4096 (bit); sigalg: RSA-SHA256
   v:NotBefore: Nov  9 12:56:23 2022 GMT; NotAfter: Nov  6 12:56:23 2032 GMT
-----BEGIN CERTIFICATE-----
MIIGWzCCBEOgAwIBAgIJAO0cnyfiJt3yMA0GCSqGSIb3DQEBCwUAMIGDMQswCQYD
VQQGEwJHQjEPMA0GA1UECAwGTG9uZG9uMQ8wDQYDVQQHDAZMb25kb24xGDAWBgNV
BAoMD0ZhbGxvd2ZpZWxkIExMQzESMBAGA1UEAwwJbG9jYWxob3N0MSQwIgYJKoZI
hvcNAQkBFhVmZ2YxQHN0LWFuZHJld3MuYWMudWswHhcNMjIxMTA5MTI1NjIzWhcN
MzIxMTA2MTI1NjIzWjCBgzELMAkGA1UEBhMCR0IxDzANBgNVBAgMBkxvbmRvbjEP
MA0GA1UEBwwGTG9uZG9uMRgwFgYDVQQKDA9GYWxsb3dmaWVsZCBMTEMxEjAQBgNV
BAMMCWxvY2FsaG9zdDEkMCIGCSqGSIb3DQEJARYVZmdmMUBzdC1hbmRyZXdzLmFj
LnVrMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA113eqff67ZP86Atj
wCf/BA7TUsSgF3gFx7AI4+ltkGxOspptKTVyIRxQYZghZSu/jmwB4ol/iD5Mj7cg
wHK47CcqEo7nzmb/tavLV60nhyHm2nDBvQzjF5ckqIwPIZ2jxde8x91rccgS6DCS
fuqkIe4DmAZIIt79vfm60kurBFfZXf8zfhL2sKBx2gtbIxoGRc/AUasKYHno12Xl
LbeG92em32G6KeV+oYWcIaVZnoqF30g8q0I3BOyGh+DMyYoWn/+02HAw/dsYgD79
bix+YyIUsJ0xeKBKMUuV+gIflf931UkN7wWjXKV+Np6epVt6Uk18qdF2Ewv8gdXX
Avm729Uy/hVOt9k4nmbdSjMl7Rv3sIApnJ6MYvIU/geY4FBf2ByvVWJIpcgw1rYP
SP84KZAguHzh+9pyQe9NpXEcxnuTavHnj+CsrXCWEougBHx+aaX3BpyOxr3tNb0K
/Pvm6zBsPNX1TCN7Ptru8SVMiM223MmLdnBFNITLn5cjEaI4dwVyllOi9QZVeUtD
MkDl7whMmC6ooAqlIlN2RyceJtWdvYY2zFf0dTUT1U2ldZV/0L1WtTgxEqklbYdt
jy6AWp7WFDPbWXKWtOO2f4tpNDwvLAaJnRhLjnYi+1WcvMFiiAl4kji/XPklhwOQ
d2YsJ4tsm1jCPdtQVHleRhCv9GMCAwEAAaOBzzCBzDAdBgNVHQ4EFgQUqG62s6Pc
JUOYQYJQrIXuHkD7K68wHwYDVR0jBBgwFoAUqG62s6PcJUOYQYJQrIXuHkD7K68w
CQYDVR0TBAIwADALBgNVHQ8EBAMCAqQwRAYDVR0RBD0wO4IJbG9jYWxob3N0ghVs
b2NhbGhvc3QubG9jYWxkb21haW6CCTEyNy4wLjAuMYIDOjoxggdmZTgwOjoxMCwG
CWCGSAGG+EIBDQQfFh1PcGVuU1NMIEdlbmVyYXRlZCBDZXJ0aWZpY2F0ZTANBgkq
hkiG9w0BAQsFAAOCAgEAIsBrjyFQkn5a1eaWJJrkROe3udgSrh66qRKO1OuxWDAO
Q29oZ7fMUdDo/KjnAoA9JkxFV8JrUUU96V7SYbRiO+yX8I9Xs6rJW/4Ge62jKUbu
178nDY/0kqEcxTsaSBMYJhlTG60G8VjqnKAsCazi2/E/QyTT1F8+Shx30aAoKPs5
9iIeCu/C4Z9RunQN+DPSE+XOrCj5lLRAdSEjFUO/HB9qvMyXYnUwgj06b+++9u9a
/YCRrmAvb8yOkP+AGwOzCkkCmGxnm0ibXNekd3RoyJTsK5UqTypiy4HlZR06Gy/s
5Xs+CvGpfVKgQATHdptfifmRY2IfCKbk8/O+HZXWaAt4NsC0NpfGS9RBZabwnVQz
eLjPL4BOIba/BK4lELFIcoJKLPmEx0zBLQi7aVfZDPdVXCYCFeEn74jIh5+BgQKv
dH1HzzLuPua14So1+4omCL2GUuq0uroZX0uAj0E3qyO8F/5Vb7H9tt90lNmxKSpl
2Q3zVa+6Q2piQh+rtkLxeUe63OA3V+ueN2ysGBBfRO2jQcTyeaclkpe814Be0XNS
UdQ2m2cwVmhGmU/ciElG+d61lsWHEmRnQKEnfRQq4zt38xI3jil5ZNIFHyXNm7Dh
yi+3QGGzL/M2WJJQEql+Xzb3mfQRIULVBvudd939pHLKZBzXeUradEKMmwUbVak=
-----END CERTIFICATE-----
---
Server certificate
subject=C = GB, ST = London, L = London, O = My Company, CN = localhost, emailAddress = email@email.com
issuer=C = GB, ST = London, L = London, O = My Company, CN = localhost, emailAddress = email@email.com
---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 2443 bytes and written 377 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 4096 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
---
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
    Session-ID: 95905A336CFD8A6CDF6F2DDF6783F065AC2F0B951FF1F4CE8816E2B24A3B59F4
    Session-ID-ctx: 
    Resumption PSK: BC237075C6802E9741D647C6FDE3A37BBA54B1EA27EE0ACCD076ECF7327E85C3A9FA0EF619D654A65BFCB67B6BE6B473
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 7200 (seconds)
    TLS session ticket:
    0000 - 55 7c 92 4c 30 55 bf 0a-f2 53 86 6d ee e1 04 88   U|.L0U...S.m....
    0010 - 3f c7 57 02 28 f5 0c 8b-77 c5 b5 ce 7d 2b 35 37   ?.W.(...w...}+57
    0020 - 94 19 e0 26 4e 03 9b b0-b5 aa 3d 07 19 d7 54 62   ...&N.....=...Tb
    0030 - 97 82 0e e5 6d 3e 36 74-67 33 14 19 3d ea 4a e8   ....m>6tg3..=.J.
    0040 - c5 65 57 92 00 e9 f8 08-15 b0 53 68 fe cd 6d ec   .eW.......Sh..m.
    0050 - ef 71 b2 d3 0e bf 9a 56-2a bf 87 9c 5e 77 40 8b   .q.....V*...^w@.
    0060 - 9c cd d5 e2 57 fe ff 2b-7a b6 80 33 a0 c5 40 ba   ....W..+z..3..@.
    0070 - 3a 2f 6d a7 2c a7 a8 c7-e5 93 8a 4d 9e ef 33 b7   :/m.,......M..3.
    0080 - 59 2c b5 64 f8 34 d5 7f-da 3d 89 28 3a ea a2 b7   Y,.d.4...=.(:...
    0090 - e9 24 ea d5 16 7b 5f 96-86 75 4d 9c 05 4b d3 21   .$...{_..uM..K.!
    00a0 - 38 b9 7a ab 6d 5a 24 7e-a2 d0 3c 6d 5e 96 68 5e   8.z.mZ$~..<m^.h^
    00b0 - f8 2b cd 2d 47 ee d5 cd-2f b3 92 1f 0e 25 b9 42   .+.-G.../....%.B
    00c0 - ef 75 a0 aa 66 c6 32 ed-5e fe 4a 7d ea 27 4a 05   .u..f.2.^.J}.'J.

    Start Time: 1668023136
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
    Max Early Data: 0
---
read R BLOCK
---
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
    Session-ID: 8A8C5C488885A20AFF95360749D1E143FB77F1CE4B7F2BC2699B1D05A0EB13FC
    Session-ID-ctx: 
    Resumption PSK: 65CECFDBD995CB56C67439AFE5BAB565E6AE9B54BFB3178225ED504CBEFFF60C6CB9C51E5E8B9C6B7B190E4FFE5B47E3
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 7200 (seconds)
    TLS session ticket:
    0000 - 55 7c 92 4c 30 55 bf 0a-f2 53 86 6d ee e1 04 88   U|.L0U...S.m....
    0010 - c6 86 9d 6a 17 e6 58 56-28 53 71 be 1f b1 2f e5   ...j..XV(Sq.../.
    0020 - 01 14 b4 f4 3a 84 cd dc-8d ea 88 df 85 ab 4f eb   ....:.........O.
    0030 - 9b 47 ee c9 c1 bc 08 b8-1d b4 e3 65 33 5a b2 bb   .G.........e3Z..
    0040 - ad e2 b7 7d 40 26 ef 79-33 49 de 8c 5d 38 6f 9d   ...}@&.y3I..]8o.
    0050 - 2a 0e 9b 9e 8f 6d 05 05-f2 00 05 71 42 ec 2c 96   *....m.....qB.,.
    0060 - b2 ad 1c c9 9b 4d 18 29-d4 6e ec 60 c1 50 4c 26   .....M.).n.`.PL&
    0070 - 34 77 6a 9e 07 73 09 be-84 30 ee fa e8 3a 1c 2b   4wj..s...0...:.+
    0080 - da 67 ed ba 5c bd 99 55-ff 99 1d 61 e4 fd 0f 7b   .g..\..U...a...{
    0090 - 05 c3 74 1d 6a 4c aa 80-33 ef 33 87 c1 f4 61 58   ..t.jL..3.3...aX
    00a0 - e5 eb 68 1e ab 08 16 3f-00 19 ee 28 71 07 75 22   ..h....?...(q.u"
    00b0 - 10 9e a5 b9 77 a8 c7 2e-ba 99 7a e9 1b 7d 79 6e   ....w.....z..}yn
    00c0 - c4 d8 a8 60 32 db f0 b3-c3 1d 60 0a a4 5f fd ee   ...`2.....`.._..

    Start Time: 1668023136
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
    Max Early Data: 0
---
read R BLOCK
```

I have tried changing the `SSL_CTX_load_verify_locations` to explicitly pick up the path to my certificates folder, and again verified that this is the correct location, but no matter what I try it fails with the same error:

```
const char* path = getenv(X509_get_default_cert_dir_env());
if(!path) {path = X509_get_default_cert_dir();}

printf(path, "\n");

printf("Success: %i\n", SSL_CTX_load_verify_locations(ctx, NULL, path));

SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, NULL);
```

Does anyone have any insight into this issue? I'm not sure whether it is a bug or if there is a subtle problem with a certificate which seems strange given s_client works.

**Comment by exitc0de** [comment]
I believe the culprit is:

```
/* Set hostname for SNI */
SSL_set_tlsext_host_name(ssl, rem_server_ip);
/* Configure server hostname check */
SSL_set1_host(ssl, rem_server_ip);
```

Removing these checks appears to fix the issue. I'm not sure why, because my cert includes both localhost and 127.0.0.1 in its SANs...

**Comment by mattcaswell** [comment]
Can you show us the contents of the localhost.conf that you used and the output of this command for the cert you generated:

````
openssl x509 -in cert.pem -noout -text
````

**Comment by exitc0de** [comment]
> Can you show us the contents of the localhost.conf that you used and the output of this command for the cert you generated:
> 
> ```
> openssl x509 -in cert.pem -noout -text
> ```

Sure, here's the certificate:

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 17085706079983558130 (0xed1c9f27e226ddf2)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: C=GB, ST=London, L=London, O=My Company, CN=localhost/emailAddress=email@email.com
        Validity
            Not Before: Nov  9 12:56:23 2022 GMT
            Not After : Nov  6 12:56:23 2032 GMT
        Subject: C=GB, ST=London, L=London, O=My Company, CN=localhost/emailAddress=email@email.com
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (4096 bit)
                Modulus:
                    00:d7:5d:de:a9:f7:fa:ed:93:fc:e8:0b:63:c0:27:
                    ff:04:0e:d3:52:c4:a0:17:78:05:c7:b0:08:e3:e9:
                    6d:90:6c:4e:b2:9a:6d:29:35:72:21:1c:50:61:98:
                    21:65:2b:bf:8e:6c:01:e2:89:7f:88:3e:4c:8f:b7:
                    20:c0:72:b8:ec:27:2a:12:8e:e7:ce:66:ff:b5:ab:
                    cb:57:ad:27:87:21:e6:da:70:c1:bd:0c:e3:17:97:
                    24:a8:8c:0f:21:9d:a3:c5:d7:bc:c7:dd:6b:71:c8:
                    12:e8:30:92:7e:ea:a4:21:ee:03:98:06:48:22:de:
                    fd:bd:f9:ba:d2:4b:ab:04:57:d9:5d:ff:33:7e:12:
                    f6:b0:a0:71:da:0b:5b:23:1a:06:45:cf:c0:51:ab:
                    0a:60:79:e8:d7:65:e5:2d:b7:86:f7:67:a6:df:61:
                    ba:29:e5:7e:a1:85:9c:21:a5:59:9e:8a:85:df:48:
                    3c:ab:42:37:04:ec:86:87:e0:cc:c9:8a:16:9f:ff:
                    b4:d8:70:30:fd:db:18:80:3e:fd:6e:2c:7e:63:22:
                    14:b0:9d:31:78:a0:4a:31:4b:95:fa:02:1f:95:ff:
                    77:d5:49:0d:ef:05:a3:5c:a5:7e:36:9e:9e:a5:5b:
                    7a:52:4d:7c:a9:d1:76:13:0b:fc:81:d5:d7:02:f9:
                    bb:db:d5:32:fe:15:4e:b7:d9:38:9e:66:dd:4a:33:
                    25:ed:1b:f7:b0:80:29:9c:9e:8c:62:f2:14:fe:07:
                    98:e0:50:5f:d8:1c:af:55:62:48:a5:c8:30:d6:b6:
                    0f:48:ff:38:29:90:20:b8:7c:e1:fb:da:72:41:ef:
                    4d:a5:71:1c:c6:7b:93:6a:f1:e7:8f:e0:ac:ad:70:
                    96:12:8b:a0:04:7c:7e:69:a5:f7:06:9c:8e:c6:bd:
                    ed:35:bd:0a:fc:fb:e6:eb:30:6c:3c:d5:f5:4c:23:
                    7b:3e:da:ee:f1:25:4c:88:cd:b6:dc:c9:8b:76:70:
                    45:34:84:cb:9f:97:23:11:a2:38:77:05:72:96:53:
                    a2:f5:06:55:79:4b:43:32:40:e5:ef:08:4c:98:2e:
                    a8:a0:0a:a5:22:53:76:47:27:1e:26:d5:9d:bd:86:
                    36:cc:57:f4:75:35:13:d5:4d:a5:75:95:7f:d0:bd:
                    56:b5:38:31:12:a9:25:6d:87:6d:8f:2e:80:5a:9e:
                    d6:14:33:db:59:72:96:b4:e3:b6:7f:8b:69:34:3c:
                    2f:2c:06:89:9d:18:4b:8e:76:22:fb:55:9c:bc:c1:
                    62:88:09:78:92:38:bf:5c:f9:25:87:03:90:77:66:
                    2c:27:8b:6c:9b:58:c2:3d:db:50:54:79:5e:46:10:
                    af:f4:63
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Subject Key Identifier: 
                A8:6E:B6:B3:A3:DC:25:43:98:41:82:50:AC:85:EE:1E:40:FB:2B:AF
            X509v3 Authority Key Identifier: 
                keyid:A8:6E:B6:B3:A3:DC:25:43:98:41:82:50:AC:85:EE:1E:40:FB:2B:AF

            X509v3 Basic Constraints: 
                CA:FALSE
            X509v3 Key Usage: 
                Digital Signature, Key Encipherment, Certificate Sign
            X509v3 Subject Alternative Name: 
                DNS:localhost, DNS:localhost.localdomain, DNS:127.0.0.1, DNS:::1, DNS:fe80::1
            Netscape Comment: 
                OpenSSL Generated Certificate
    Signature Algorithm: sha256WithRSAEncryption
         22:c0:6b:8f:21:50:92:7e:5a:d5:e6:96:24:9a:e4:44:e7:b7:
         b9:d8:12:ae:1e:ba:a9:12:8e:d4:eb:b1:58:30:0e:43:6f:68:
         67:b7:cc:51:d0:e8:fc:a8:e7:02:80:3d:26:4c:45:57:c2:6b:
         51:45:3d:e9:5e:d2:61:b4:62:3b:ec:97:f0:8f:57:b3:aa:c9:
         5b:fe:06:7b:ad:a3:29:46:ee:d7:bf:27:0d:8f:f4:92:a1:1c:
         c5:3b:1a:48:13:18:26:19:53:1b:ad:06:f1:58:ea:9c:a0:2c:
         09:ac:e2:db:f1:3f:43:24:d3:d4:5f:3e:4a:1c:77:d1:a0:28:
         28:fb:39:f6:22:1e:0a:ef:c2:e1:9f:51:ba:74:0d:f8:33:d2:
         13:e5:ce:ac:28:f9:94:b4:40:75:21:23:15:43:bf:1c:1f:6a:
         bc:cc:97:62:75:30:82:3d:3a:6f:ef:be:f6:ef:5a:fd:80:91:
         ae:60:2f:6f:cc:8e:90:ff:80:1b:03:b3:0a:49:02:98:6c:67:
         9b:48:9b:5c:d7:a4:77:74:68:c8:94:ec:2b:95:2a:4f:2a:62:
         cb:81:e5:65:1d:3a:1b:2f:ec:e5:7b:3e:0a:f1:a9:7d:52:a0:
         40:04:c7:76:9b:5f:89:f9:91:63:62:1f:08:a6:e4:f3:f3:be:
         1d:95:d6:68:0b:78:36:c0:b4:36:97:c6:4b:d4:41:65:a6:f0:
         9d:54:33:78:b8:cf:2f:80:4e:21:b6:bf:04:ae:25:10:b1:48:
         72:82:4a:2c:f9:84:c7:4c:c1:2d:08:bb:69:57:d9:0c:f7:55:
         5c:26:02:15:e1:27:ef:88:c8:87:9f:81:81:02:af:74:7d:47:
         cf:32:ee:3e:e6:b5:e1:2a:35:fb:8a:26:08:bd:86:52:ea:b4:
         ba:ba:19:5f:4b:80:8f:41:37:ab:23:bc:17:fe:55:6f:b1:fd:
         b6:df:74:94:d9:b1:29:2a:65:d9:0d:f3:55:af:ba:43:6a:62:
         42:1f:ab:b6:42:f1:79:47:ba:dc:e0:37:57:eb:9e:37:6c:ac:
         18:10:5f:44:ed:a3:41:c4:f2:79:a7:25:92:97:bc:d7:80:5e:
         d1:73:52:51:d4:36:9b:67:30:56:68:46:99:4f:dc:88:49:46:
         f9:de:b5:96:c5:87:12:64:67:40:a1:27:7d:14:2a:e3:3b:77:
         f3:12:37:8e:29:79:64:d2:05:1f:25:cd:9b:b0:e1:ca:2f:b7:
         40:61:b3:2f:f3:36:58:92:50:12:a9:7e:5f:36:f7:99:f4:11:
         21:42:d5:06:fb:9d:77:dd:fd:a4:72:ca:64:1c:d7:79:4a:da:
         74:42:8c:9b:05:1b:55:a9
```

localhost.conf:

```
# Self Signed (note the addition of -x509):
#     openssl req -config example-com.conf -new -x509 -sha256 -newkey rsa:2048 -nodes -keyout example-com.key.pem -days 365 -out example-com.cert.pem
# Signing Request (note the lack of -x509):
#     openssl req -config example-com.conf -new -newkey rsa:2048 -nodes -keyout example-com.key.pem -days 365 -out example-com.req.pem
# Print it:
#     openssl x509 -in example-com.cert.pem -text -noout
#     openssl req -in example-com.req.pem -text -noout

[ req ]
default_bits        = 4096
default_keyfile     = server-key.pem
distinguished_name  = subject
req_extensions      = req_ext
x509_extensions     = x509_ext
string_mask         = utf8only

# The Subject DN can be formed using X501 or RFC 4514 (see RFC 4519 for a description).
#   It's sort of a mashup. For example, RFC 4514 does not provide emailAddress.
[ subject ]
countryName         = Country Name (2 letter code)
countryName_default     = GB

stateOrProvinceName     = State or Province Name (full name)
stateOrProvinceName_default = London

localityName            = Locality Name (eg, city)
localityName_default        = London

organizationName         = Organization Name (eg, company)
organizationName_default    = My Company

# Use a friendly name here because it's presented to the user. The server's DNS
#   names are placed in Subject Alternate Names. Plus, DNS names here is deprecated
#   by both IETF and CA/Browser Forums. If you place a DNS name here, then you 
#   must include the DNS name in the SAN too (otherwise, Chrome and others that
#   strictly follow the CA/Browser Baseline Requirements will fail).
commonName          = Common Name (e.g. server FQDN or YOUR name)
commonName_default      = localhost

emailAddress            = Email Address
emailAddress_default        = email@email.com

# Section x509_ext is used when generating a self-signed certificate. I.e., openssl req -x509 ...
[ x509_ext ]

subjectKeyIdentifier        = hash
authorityKeyIdentifier  = keyid,issuer

#  If RSA Key Transport bothers you, then remove keyEncipherment. TLS 1.3 is removing RSA
#  Key Transport in favor of exchanges with Forward Secrecy, like DHE and ECDHE.
basicConstraints        = CA:FALSE
keyUsage            = digitalSignature, keyEncipherment, keyCertSign
subjectAltName          = @alternate_names
nsComment           = "OpenSSL Generated Certificate"

# RFC 5280, Section 4.2.1.12 makes EKU optional
# CA/Browser Baseline Requirements, Appendix (B)(3)(G) makes me confused
# extendedKeyUsage  = serverAuth, clientAuth

# Section req_ext is used when generating a certificate signing request. I.e., openssl req ...
[ req_ext ]

subjectKeyIdentifier        = hash

basicConstraints        = CA:FALSE
keyUsage            = digitalSignature, keyEncipherment, keyCertSign
subjectAltName          = @alternate_names
nsComment           = "OpenSSL Generated Certificate"

# RFC 5280, Section 4.2.1.12 makes EKU optional
# CA/Browser Baseline Requirements, Appendix (B)(3)(G) makes me confused
# extendedKeyUsage  = serverAuth, clientAuth

[ alternate_names ]

DNS.1       = localhost
DNS.2       = localhost.localdomain
DNS.3       = 127.0.0.1
# IPv6 localhost
DNS.4       = ::1
DNS.5       = fe80::1
```

**Comment by Leseratte10** [comment]
This here seems to be wrong: 

```
[ alternate_names ]

DNS.1       = localhost
DNS.2       = localhost.localdomain
DNS.3       = 127.0.0.1
# IPv6 localhost
DNS.4       = ::1
DNS.5       = fe80::1
```

"DNS" is only for actual DNS names. You need to use "IP" for IP addresses. Try this one and see if that helps: 


```
[ alternate_names ]

DNS.1       = localhost
DNS.2       = localhost.localdomain
IP.1       = 127.0.0.1
IP.2       = ::1
IP.3       = fe80::1
```

**Comment by nhorman** [comment]
marking as inactive, will be closed at the completion of 3.4 dev without further input
